gems/sorbet/lib/serialize.rb

```diff
@@ -253,7 +253,7 @@ def serialize_method(method, static=false, with_sig: true)
 
   def valid_class_name(name)
     name.split("::").each do |piece|
-      return false if piece[0].upcase != piece[0]
+      return false unless piece =~ /^[[:upper:]][[:word:]]*$/
```

Comment:
I couldn't find any authoritative reference on the question of what non-ASCII characters are legal for class names in Ruby. It seems that maybe anything goes when it comes to Unicode; but since these character classes seem to do the right thing in ASCII I'm hoping the spirit of them is consistent in other character sets, and reasonable to use here.

We could also just add a special case to exclude `#`, but there are probably other things that would cause syntax problems (although it might be harder to wind up with them).

Reply:
Do you think you could change this so that we filter out the things we know aren't ok, rather than attempt to do a complete validation, and potentially reject things that are fine?

For example, I think

- return false if spaces
- return false if punctuation
- return false if doesn't start with uppercase

is similar in spirit to what you've written here, but has a lower chance of breaking valid usage.

Reply:
So, I tested this, and it works on my codebase:
```ruby
      return false unless piece =~ /^[[:upper:]]/
      return false if piece =~ /[[:space:][:cntrl:]]/
      return false if piece =~ /[[:punct:]&&[^_]]/
```

But I also noticed that `:word:` is already used in [valid_method_name](https://github.com/sorbet/sorbet/blob/a630ef3d431c2b1c4a1ff979abcc0f78a6e8af46/gems/sorbet/lib/serialize.rb#L202), so it seems like it would make sense to apply the same rule(s) in both places. Thoughts?

Reply:
That regex is already too restrictive. `def !(); end` and `def ==(); end` are both valid method definitions that regex rejects.

More generally: all this runtime reflection code is incredibly brittle and poorly tested. I'd rather not change things that aren't actively causing problems. Let's just leave it at what you've implemented and call it done.

Reply:
:+1: updated the PR